<a href="https://colab.research.google.com/github/VintaBytes/Ciencia-de-datos/blob/main/libros/ciencia-de-datos-con-python/volumen-02/cuadernos/capitulo-028-agrupar-datos-con-groupby.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Abrir en Colab"/></a>

# Capítulo 28 — Agrupar datos con `groupby()`

## De las tablas generales a los resúmenes por grupo

En los capítulos anteriores construimos resúmenes generales del dataset.

Calculamos indicadores numéricos, analizamos frecuencias categóricas y organizamos resultados en tablas de resumen. Esas herramientas nos permitieron responder preguntas como:

- ¿cuál es el promedio general de las cuentas?
- ¿cuántos registros hay para cada día?
- ¿qué porcentaje del dataset representa cada categoría?
- ¿cómo presentar indicadores de forma más clara?

Ahora vamos a avanzar hacia una pregunta más analítica:

**¿Cómo cambian esos indicadores según un grupo?**

Por ejemplo, no solo queremos saber cuál es el promedio general de las cuentas. También podríamos preguntarnos:

**¿Cuál es el promedio de cuenta para cada día?**

Para responder ese tipo de preguntas necesitamos agrupar datos.

## Qué significa agrupar datos

Agrupar datos significa separar un dataset en partes según los valores de una columna y luego calcular algún resumen dentro de cada parte.

Por ejemplo, si agrupamos por la columna `day`, Pandas separa los registros según el día:

- registros de `Thur`;
- registros de `Fri`;
- registros de `Sat`;
- registros de `Sun`.

Luego podemos calcular un indicador para cada grupo, como el promedio de cuenta, el total facturado o la cantidad de registros.

En Pandas, una de las herramientas principales para hacer esto es `groupby()`.

## Cargar el dataset de trabajo

Vamos a seguir usando el dataset `tips` de Seaborn.

Esto nos permite continuar con el mismo contexto y concentrarnos en una herramienta nueva: `groupby()`.

In [29]:
import pandas as pd
import seaborn as sns

df = sns.load_dataset("tips")

df.head()

,total_bill,tip,sex,smoker,day,time,size
0,16.99,1.01,Female,No,Sun,Dinner,2
1,10.34,1.66,Male,No,Sun,Dinner,3
2,21.01,3.50,Male,No,Sun,Dinner,3
3,23.68,3.31,Male,No,Sun,Dinner,2
4,24.59,3.61,Female,No,Sun,Dinner,4


La tabla muestra las primeras filas del dataset.

Seguimos trabajando con las mismas variables de los capítulos anteriores. En este capítulo vamos a usar especialmente:

- `day`, para formar grupos según el día;
- `time`, para formar grupos según el horario;
- `total_bill`, para calcular resúmenes sobre el total de las cuentas;
- `tip`, para calcular resúmenes sobre las propinas.

La idea central será combinar una columna categórica con una columna numérica.

## Una primera pregunta por grupo

Hasta ahora sabemos calcular el promedio general de una columna.

Por ejemplo, podemos calcular el promedio general de `total_bill`. Pero ahora queremos hacer una pregunta más específica:

**¿Cuál es el promedio de cuenta para cada día?**

Para responderla, necesitamos agrupar los datos por `day` y calcular el promedio de `total_bill` dentro de cada grupo.

In [30]:
df.groupby("day", observed=True)["total_bill"].mean()

,total_bill
day,
Thur,17.682742
Fri,17.151579
Sat,20.441379
Sun,21.410000


El resultado muestra el promedio de `total_bill` para cada día.

Ahora ya no tenemos un único promedio general, sino un promedio separado por grupo:

- `Thur`: aproximadamente **17.68**.
- `Fri`: aproximadamente **17.15**.
- `Sat`: aproximadamente **20.44**.
- `Sun`: aproximadamente **21.41**.

Según estos datos, el promedio de cuenta más alto aparece el domingo (`Sun`), seguido por el sábado (`Sat`). Los promedios más bajos aparecen jueves (`Thur`) y viernes (`Fri`).

Este es el cambio central que introduce `groupby()`: nos permite calcular indicadores separados para cada categoría.

En esta celda usamos `observed=True` porque `day` es una columna categórica.

Este parámetro le indica a Pandas que trabaje con las categorías observadas en los datos. En este caso no cambia el resultado, pero evita un aviso relacionado con cambios futuros en Pandas.

In [31]:
promedio_cuenta_por_dia = df.groupby("day", observed=True)["total_bill"].mean()

promedio_cuenta_por_dia = promedio_cuenta_por_dia.round(2)

promedio_cuenta_por_dia

,total_bill
day,
Thur,17.68
Fri,17.15
Sat,20.44
Sun,21.41


Ahora el resultado es más fácil de leer.

El promedio de cuenta por día queda así:

- `Thur`: **17.68**
- `Fri`: **17.15**
- `Sat`: **20.44**
- `Sun`: **21.41**

Esta tabla muestra que el promedio de cuenta es mayor durante el domingo y el sábado que durante el jueves y el viernes.

A diferencia de los resúmenes generales, este resultado ya compara grupos dentro del dataset. La columna `day` define los grupos y la columna `total_bill` aporta los valores numéricos que se resumen.

## Leer la estructura de un `groupby()`

La expresión que usamos puede leerse en tres partes:

```python
df.groupby("day", observed=True)["total_bill"].mean()
```

Primero, `groupby("day")` agrupa las filas según el valor de la columna `day`.

Luego, `["total_bill"]` indica qué columna numérica queremos resumir.

Finalmente, `.mean()` calcula el promedio dentro de cada grupo.

En palabras simples: agrupamos por día y calculamos el promedio de cuenta para cada día.

## Cambiar el indicador calculado

Una vez que tenemos clara la idea de agrupar, podemos cambiar el indicador.

En lugar de calcular el promedio de cuenta por día, podemos calcular el total facturado por día.

La agrupación será la misma: `day`. Lo que cambia es el cálculo final.

In [32]:
total_facturado_por_dia = df.groupby("day", observed=True)["total_bill"].sum()

total_facturado_por_dia = total_facturado_por_dia.round(2)

total_facturado_por_dia

,total_bill
day,
Thur,1096.33
Fri,325.88
Sat,1778.40
Sun,1627.16


Ahora calculamos el **total facturado por día**.

El resultado muestra que:

- `Sat` tiene el total facturado más alto: **1778.40**.
- `Sun` aparece en segundo lugar: **1627.16**.
- `Thur` suma **1096.33**.
- `Fri` tiene el total más bajo: **325.88**.

Este resultado no debe leerse igual que el promedio anterior.

El promedio de cuenta nos dice cuánto se consume, en promedio, por registro. El total facturado depende tanto del valor de las cuentas como de la cantidad de registros que hay para cada día.

Por eso, un día puede tener mayor facturación total no solo porque sus cuentas sean más altas, sino también porque tiene más cuentas registradas.

## Calcular la cantidad de registros por grupo

Para interpretar mejor un total, también conviene saber cuántos registros hay en cada grupo.

Vamos a contar cuántas cuentas hay registradas para cada día usando `groupby()`.

In [33]:
cantidad_cuentas_por_dia = df.groupby("day", observed=True)["total_bill"].count()

cantidad_cuentas_por_dia

,total_bill
day,
Thur,62
Fri,19
Sat,87
Sun,76


La tabla muestra cuántas cuentas hay registradas para cada día:

- `Thur`: **62** cuentas.
- `Fri`: **19** cuentas.
- `Sat`: **87** cuentas.
- `Sun`: **76** cuentas.

Esta información ayuda a interpretar el total facturado por día.

Por ejemplo, `Sat` tiene el total facturado más alto, pero también es el día con más cuentas registradas. En cambio, `Sun` tiene un promedio de cuenta más alto que `Sat`, aunque su total facturado es menor porque tiene menos registros.

Esto muestra por qué conviene mirar más de un indicador antes de sacar conclusiones.

## Combinar varios resultados por grupo

Hasta ahora calculamos tres resultados separados por día:

- promedio de cuenta;
- total facturado;
- cantidad de cuentas.

Para analizar mejor los grupos, puede ser útil reunir esos resultados en una sola tabla.

In [34]:
resumen_por_dia = pd.DataFrame({
    "cantidad_cuentas": cantidad_cuentas_por_dia,
    "total_facturado": total_facturado_por_dia,
    "promedio_cuenta": promedio_cuenta_por_dia
})

resumen_por_dia

,cantidad_cuentas,total_facturado,promedio_cuenta
day,,,
Thur,62,1096.33,17.68
Fri,19,325.88,17.15
Sat,87,1778.40,20.44
Sun,76,1627.16,21.41


Ahora tenemos una tabla de resumen por día.

Cada fila representa un grupo definido por la columna `day`, y cada columna muestra un indicador distinto:

- `cantidad_cuentas`: cantidad de registros para ese día.
- `total_facturado`: suma de `total_bill` para ese día.
- `promedio_cuenta`: promedio de `total_bill` para ese día.

Esta tabla permite leer mejor las diferencias entre grupos.

Por ejemplo, `Sat` tiene la mayor cantidad de cuentas y el mayor total facturado. Sin embargo, `Sun` tiene el promedio de cuenta más alto. Esto muestra que el total facturado y el promedio de cuenta no responden exactamente la misma pregunta.

## Calcular varios indicadores con `agg()`

La tabla anterior fue construida combinando varios resultados separados.

Pandas también permite calcular varios indicadores dentro de una misma agrupación usando `.agg()`.

La palabra `agg` viene de “aggregate”, que podemos entender como “agregar” o “resumir”. En este contexto significa aplicar una o varias funciones de resumen sobre cada grupo.

In [35]:
resumen_por_dia_agg = df.groupby("day", observed=True)["total_bill"].agg(
    cantidad_cuentas="count",
    total_facturado="sum",
    promedio_cuenta="mean"
)

resumen_por_dia_agg = resumen_por_dia_agg.round(2)

resumen_por_dia_agg

,cantidad_cuentas,total_facturado,promedio_cuenta
day,,,
Thur,62,1096.33,17.68
Fri,19,325.88,17.15
Sat,87,1778.40,20.44
Sun,76,1627.16,21.41


El resultado es el mismo que habíamos construido manualmente, pero ahora lo obtuvimos con una sola agrupación y una sola llamada a `.agg()`.

Esto tiene una ventaja importante: el código queda más compacto y expresa mejor la intención del análisis.

La instrucción puede leerse así:

- agrupar por `day`;
- tomar la columna `total_bill`;
- calcular cantidad, suma y promedio para cada grupo;
- asignar nombres claros a cada indicador.

Cuando necesitamos calcular varios resúmenes sobre una misma columna, `.agg()` suele ser una forma más ordenada de hacerlo.

## Convertir el índice en una columna

Después de usar `groupby()`, la columna usada para agrupar suele quedar como índice del resultado.

En este caso, `day` aparece como índice de la tabla.

Muchas veces conviene convertir ese índice nuevamente en una columna común, especialmente si queremos seguir trabajando con el resultado como un `DataFrame`.

In [36]:
resumen_por_dia_agg = resumen_por_dia_agg.reset_index()

resumen_por_dia_agg

,day,cantidad_cuentas,total_facturado,promedio_cuenta
0,Thur,62,1096.33,17.68
1,Fri,19,325.88,17.15
2,Sat,87,1778.40,20.44
3,Sun,76,1627.16,21.41


Ahora `day` vuelve a aparecer como una columna común del `DataFrame`.

Esto puede resultar más cómodo para leer, ordenar, filtrar o usar la tabla en pasos posteriores.

El contenido no cambió: seguimos teniendo los mismos grupos y los mismos indicadores. Lo que cambió es la forma de presentar la tabla.

## Agrupar por otra columna

La misma lógica puede aplicarse a otra variable categórica.

Ahora vamos a agrupar por `time`, que distingue entre almuerzo (`Lunch`) y cena (`Dinner`), y vamos a calcular los mismos indicadores sobre `total_bill`.

In [37]:
resumen_por_horario = df.groupby("time", observed=True)["total_bill"].agg(
    cantidad_cuentas="count",
    total_facturado="sum",
    promedio_cuenta="mean"
)

resumen_por_horario = resumen_por_horario.round(2).reset_index()

resumen_por_horario

,time,cantidad_cuentas,total_facturado,promedio_cuenta
0,Lunch,68,1167.47,17.17
1,Dinner,176,3660.30,20.80


La tabla resume las cuentas según el horario.

Podemos observar que:

- `Lunch` tiene **68 cuentas**, un total facturado de **1167.47** y un promedio de cuenta de **17.17**.
- `Dinner` tiene **176 cuentas**, un total facturado de **3660.30** y un promedio de cuenta de **20.80**.

En este dataset, la cena concentra más registros, más facturación total y también un promedio de cuenta más alto.

Esta tabla muestra cómo `groupby()` permite responder preguntas más específicas que un resumen general. Ya no estamos mirando solo el total del dataset, sino comparando indicadores entre grupos.

## Agrupar por más de una columna

Hasta ahora agrupamos por una sola columna: primero por `day` y luego por `time`.

Pero también podemos agrupar usando más de una columna. Esto permite responder preguntas más detalladas.

Por ejemplo:

**¿Cuántas cuentas hay para cada combinación de día y horario?**

Para responder esta pregunta, podemos agrupar por `day` y `time` al mismo tiempo.

In [38]:
cuentas_por_dia_y_horario = df.groupby(["day", "time"], observed=True)["total_bill"].count()

cuentas_por_dia_y_horario

day   time  
Thur  Lunch     61
      Dinner     1
Fri   Lunch      7
      Dinner    12
Sat   Dinner    87
Sun   Dinner    76
Name: total_bill, dtype: int64

El resultado muestra la cantidad de cuentas para cada combinación de `day` y `time`.

Ahora los grupos no se forman solo por día, sino por la combinación entre día y horario.

Podemos leer algunos resultados:

- `Thur` tiene **61 cuentas** en `Lunch` y **1 cuenta** en `Dinner`.
- `Fri` tiene **7 cuentas** en `Lunch` y **12 cuentas** en `Dinner`.
- `Sat` tiene **87 cuentas** en `Dinner`.
- `Sun` tiene **76 cuentas** en `Dinner`.

Esto permite responder una pregunta más detallada que antes. No solo sabemos cuántas cuentas hay por día o por horario, sino cuántas hay para cada combinación de ambas variables.

También aparece una diferencia en la forma de la salida: ahora el resultado tiene dos niveles de índice, porque agrupamos por dos columnas.

## Convertir una agrupación doble en tabla

Cuando agrupamos por más de una columna, Pandas puede devolver un resultado con varios niveles de índice.

Para leerlo de forma más parecida a una tabla común, podemos usar `reset_index()`.

In [39]:
cuentas_por_dia_y_horario = cuentas_por_dia_y_horario.reset_index()

cuentas_por_dia_y_horario = cuentas_por_dia_y_horario.rename(
    columns={"total_bill": "cantidad_cuentas"}
)

cuentas_por_dia_y_horario

,day,time,cantidad_cuentas
0,Thur,Lunch,61
1,Thur,Dinner,1
2,Fri,Lunch,7
3,Fri,Dinner,12
4,Sat,Dinner,87
5,Sun,Dinner,76


Ahora el resultado tiene forma de tabla común.

Las columnas `day` y `time` indican el grupo al que pertenece cada fila. La columna `cantidad_cuentas` contiene la cantidad de cuentas en esa combinación de día y horario.

La información es la misma que antes, pero esta presentación suele ser más cómoda para leer y para seguir trabajando.

Por ejemplo, podemos ver rápidamente que:

- `Thur` aparece principalmente en `Lunch`;
- `Fri` tiene registros tanto en `Lunch` como en `Dinner`;
- `Sat` y `Sun` aparecen solo con registros de `Dinner`.

Esta primera agrupación por más de una columna muestra que `groupby()` puede responder preguntas cada vez más específicas.

## Cierre del capítulo

En este capítulo comenzamos a trabajar con `groupby()`, una de las herramientas más importantes de Pandas para el análisis de datos.

La idea central fue agrupar filas según una o más columnas y calcular resúmenes dentro de cada grupo.

Vimos que podemos usar `groupby()` para responder preguntas como:

- ¿cuál es el promedio de cuenta por día?
- ¿cuál es el total facturado por día?
- ¿cuántas cuentas hay en cada grupo?
- ¿cómo cambian los indicadores según el horario?
- ¿cuántas cuentas hay para cada combinación de día y horario?

También vimos que no todos los indicadores responden la misma pregunta. El total facturado depende de los valores de las cuentas y de la cantidad de registros, mientras que el promedio de cuenta describe el valor promedio dentro de cada grupo.

Finalmente, usamos `.agg()` para calcular varios indicadores en una misma agrupación y `reset_index()` para convertir el resultado en una tabla más cómoda de leer.

En los próximos capítulos vamos a seguir usando agrupaciones para comparar categorías, construir tablas más completas y preparar resultados para visualizaciones.